### ISSUE TITLE
[Question] javascript port

### ISSUE THREAD

**Issue by y-lohse** [inkle/ink#88]
Hi, I haven't seen the question answered (or even asked!) yet, so i'll just go ahead.

Is there a more precise plan and timeline for a javascript port of the runtime engine?

The readme says you guys won't have time to work on it yourselves, but other sources are less definitive. You also mention that it might be too early for now because the engine is still changing, but do you have a vague idea of when it will be stable *enough*?

Thanks!

**Comment by joethephish** [comment]
Yeah, we don't know of anyone working on it right now, unfortunately.

I've been giving it some more thought though, and I think that our runtime engine is stable enough that someone could implement it using the current JSON based format. People have been getting a bit hung up on the exact details of the data format, when most of the work is in the runtime engine itself. Any other formats we come up with (such as "inkc", a minimal text-based format I've been experimenting with) will always follow the same basic structure, so would take very little time to implement.

I should perhaps write this up more "officially" somewhere, but for anyone thinking of doing a port of the runtime engine to another language, including JavaScript, I'd strongly recommend:

* Implementing *just* the runtime engine itself, not creating a new version of the parser from scratch. This should save you a lot of work since we solve a lot of problems at parse/compile time. The runtime engine itself is relatively simple, and it's helpful to work towards an interchange format that works everywhere.
* Follow / port the C# code as directly as possible as a reference implementation. One thing I could/should do to help with this is to restructure the code slightly into parts that are varying degrees of critical v.s. optional. For example the core execution loop is critical, but external functions less so when starting a port.

**Comment by y-lohse** [comment]
Right, thank you for the clarifications.

I actually looked a bit at the engine's source already. It still seems like a fair amount of work to port it, but if it's not going to change too much it looks a bit more achievable. My main worry is deciphering the C#-specific bits, as I haven't used that language in a while. Figuring out [that line](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Story.cs#L130) for example took me ages :stuck_out_tongue: 

Anyway, i might give it a shot and report back if I get somewhere with it.

**Comment by joethephish** [comment]
Cool, good luck!

**Comment by y-lohse** [comment]
Well, I got a bit consumed by that and I have something that's working :smile: . I've open sourced it [over here](https://github.com/y-lohse/jsink).

Right now it supports choices & diverts, but I think I've translated about 80% of the code, so I expect the rest to follow pretty soon. It's also quite similar to the reference code, though I left out vast portions for now (external functions are variables for example).

I do have a couple of questions, if you don't mind:

- [what does this line do?](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Story.cs#L159) It looks like an increment by a function definition :confused: 
- [why is there a break here?](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Object.cs#L137) Dosen't it stop the search after the first cycle?
- the documentation says to use [choice.choiceText](https://github.com/inkle/ink/blob/master/Documentation/RunningYourInk.md#getting-started-with-the-runtime-api), but as far as I can see `Choice`s only have a `text` property, not `choiceText`. Is that a mistake in the doc?

**Comment by joethephish** [comment]
Awesome, can't wait to take a look! To answer your questions:

* This line: `_state.variablesState.variableChangedEvent += VariableStateDidChangeEvent;` is equivalent to something like (in Node parlance) `_state.variablesState.on("variableChanged", VariableStateDidChangeEvent)`. The event is a C# construct that can take multiple listeners and is a language-level feature, hence the `+=`.

* That break is for comparing two paths like `a.b.c.d.e` and `a.b.c.x.y`. It steps up through a, b and c, then at d/x it breaks so that `lastSharedPathCompIndex` is 2, pointing to `c`.

* `choiceText` was renamed to `text`. I'll fix it in the docs, thanks!

**Comment by joethephish** [comment]
Awesome, can't wait to take a look! To answer your questions:

* This line: `_state.variablesState.variableChangedEvent += VariableStateDidChangeEvent;` is equivalent to something like (in Node parlance) `_state.variablesState.on("variableChanged", VariableStateDidChangeEvent)`. The event is a C# construct that can take multiple listeners and is a language-level feature, hence the `+=`.

* That break is for comparing two paths like `a.b.c.d.e` and `a.b.c.x.y`. It steps up through a, b and c, then at d/x it breaks so that `lastSharedPathCompIndex` is 2, pointing to `c`.

* `choiceText` was renamed to `text`. I'll fix it in the docs, thanks!

**Comment by joethephish** [comment]
I've just cloned jsink and been trying to build it. Just wanted to let you know about a hitch I just had - I've never used grunt before, so I got:

    npm install & grunt rollup
    -bash: grunt: command not found

(And I've never used it before... is it something that you could include in a `package.json`, or should you put something into the instructions to install it globally?)

But after installing grunt, I have now successfully built `ink.amd.js` and `ink.cjs.js` - woohoo!

I don't have much experience with modern JS - how would I go about building a "hello world" example where I embed it in a webpage? I assume I include the `ink.amd.js` script...  what else do I need to include to make it work?

Also, you might want to give a basic overview of compiling ink to json and then loading that?

Thanks!

**Comment by y-lohse** [comment]
First, thanks for the answer, they are very clear :)

Regarding grunt, I wasn't sure, but it looks like it does need a global installation to be used. I had it installed already  so I didn't notice. Ideally I'd like to drop that dependency altogether since the building part is fairly basic.

To actually run an ink, you can have a look in the test folder. There's a [html file](https://github.com/y-lohse/jsink/blob/master/test/simple.html) that loads a script and displays it in the console. I haven't gotten around building an actual UI around it, so there's currently no way to select choices.
I mostly used the [simple.js](https://github.com/y-lohse/jsink/blob/master/test/simple.js) file for testing. You can just use `node simple.js` and it plays the script & reads the choices from the terminal.

> Also, you might want to give a basic overview of compiling ink to json and then loading that?

Not sure what you mean :). I'm still using inklecate to compile ink files. But I agree that a more fleshed out starting guide is needed :stuck_out_tongue:

**Comment by joethephish** [comment]
Yeah, removing the grunt dependency if it's not bringing much sounds like a good plan.

Ah yes, after running `python -m SimpleHTTPServer` in the root directory and navigating to `test/simple.html` it worked great!

On your node in the readme - I wouldn't place too much importance on porting the compiler itself. While it would be kinda cool to be able to compile within a browser client, outside of that I'm not sure I see much value in it. It would be more important to put lots of polish into ink.js itself.

But yeah, in terms of the guide, that would be great but I agree it's early days. May as well wait until you've got a full implementation working before getting serious. And after that, we'd be very excited to promote it - lots of people have been waiting for a JavaScript implementation!

**Comment by y-lohse** [comment]
Actually I spent some time polishing things up a bit before going on with the implementation. I've thrown together [a browser demo](http://yannick-lohse.fr/inkjs/) and the readme is improved.

> While it would be kinda cool to be able to compile within a browser client, outside of that I'm not sure I see much value in it. It would be more important to put lots of polish into ink.js itself.

I agree, but aside from compiling in the browser it'd be quite helpful to have something else than a cli to do the compiling. We could use it to run some tests without storing the json files, and being able to compile ink files on a server without too many dependencies would be nice too. But yeah, it's really not high priority right now.

**Comment by joethephish** [comment]
Cool, looking good!

On the subject of having something other than a CLI to do the compiling - I've been working on an ink editor in my spare time, so that should be around soon. In fact, it would be pretty cool to be able to export a complete packaged website from the tool, including the ink.js runtime!

**Comment by y-lohse** [comment]
Hi, i'm nearly done porting the code — I still need to do variable observers and external functions, but that's pretty much it.
However, I've hit a bit of a rock with NativeFunctionCalls and was hoping you could point me in the right direction.i'm doing tests on [a very simple condition](https://github.com/y-lohse/inkjs/blob/master/test/stories/var.ink.json).

In my understanding, when the JSON converter hits the condition, he creates a [new NativeFunctionCall object](https://github.com/inkle/ink/blob/master/ink-engine-runtime/JsonSerialisation.cs#L162). Then a bit later, [that object gets evaluated and a call is made](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Story.cs#L907).

When the call is made, the code is using `func.numberOfParameters`… but I don't understand when this would have been set (and indeed in my code, it is never set). The only time where this property seems to be changed is when the class is internally creating the native functions… but that's not used during JSON deserialisation.

Any idea of what I coul have missed?

Thanks!

**Comment by joethephish** [comment]
Hey, glad the port is coming along well. Externals and variable observers are definitely less critical features (and also reasonably simple), so it sounds like you're in a good place! Have you got `StoryState` saving working already?

So, the `NativeFunctionCall` code is slightly odd, but what happens is:

First time a particular `NativeFunctionCall` is needed, the `GenerateNativeFunctionsIfNecessary` is run, and it creates a set of "prototype" functions (unrelated to the JavaScript sense). They're actual instances of the `NativeFunctionCall`, and are stored in a dictionary (`_nativeFunctions`) that's indexed by the unique name of the function. (The negate unary op is disambiguated from subtract by using `"~"` rather than `"-"`.) If you look through the `GenerateNativeFunctionsIfNecessary` method, it defines all of the native functions using lambda/anonymous function syntax. e.g. the definition of float multiplication is:

    AddFloatBinaryOp(Multiply, (x, y) => x * y);

In ES6 Javascript this would written in exactly the same way (`Multiply` is defined as the string `*`). Looking at the definition of `AddFloatBinaryOp`, we see where the number of parameters is defined:

    static void AddFloatBinaryOp(string name, BinaryOp<float> op)
    {
        AddOpToNativeFunc (name, 2, ValueType.Float, op);
    }

The 2 is because it's a binary operation. `AddOpToNativeFunc` simply creates the `NativeFunctionCall` object, passing in the anonymous function and the number of arguments (parameters) it takes.

So, once all the native functions *prototypes* have been generated, then when an "instance" is needed, it is simply given a name, so that it can look up its prototype when it needs to be used - **this is done in the `name` property**.

Yeah, it's slightly odd. You might wonder why we shouldn't simply return the literal prototype instance from `CallWithName` so it doesn't need to create a load of dummy instances. But it's necessary for each call instance to be unique, so that as a `Runtime.Object`, it can have its own parent and path within the story's hierarchy, rather than having the same one referenced from multiple places.

Does that make sense? Let me know if I can clarify anything further!

One last thing on an unrelated topic: I noticed in your JSON example that you had `"t?"` and `"f?"` objects representing `Runtime.Branch` objects. The branch object has recently been removed in favour of adding a simple condition to `Runtime.Divert` objects (see commit 0f8e876c67dc74783b682c1e367f10b85cfe3780). The change was partly intended to make ports easier since it's a simplification of the runtime and file format. So, apologies for actually making it more complex for you in this case!

I guess the right thing to do is to get your code fully working to match a specific point in the commit history, and then integrate fixes and changes afterwards? That's the only major change to the JSON format that I can think of right now.

**Comment by y-lohse** [comment]
Aaaah thank you!

>  so that it can look up its prototype when it needs to be used - this is done in the name property.

That's the part I missed :stuck_out_tongue: . I mean the code is there, but static functions are a bit weird, so I probably messed up something there.

>  Have you got StoryState saving working already?

Actually no, i forgot about that bit. Most of the code for that is already ported, but I haven't tried it yet. Maybe it works, for all I know!

Thanks for the head's up regarding branches. But yes, I'll first try to match the fork I've used as a reference, and patch the outstanding changes after that.

**Comment by joethephish** [comment]
Excellent, ever closer to completion! I guess a good test for the final working version will be a compiled version of [The Intercept](https://github.com/inkle/the-intercept/blob/master/Assets/Ink/TheIntercept.ink) :)

**Comment by y-lohse** [comment]
Aw man you shouldn't have said that. Of course I just *had* to  try it out, and then fix the bunch of bugs that prevented it from working.
But anyway, it mostly working now. I've updated [the demo](http://yannick-lohse.fr/inkjs/) so that the sample story is now The Intercept. I think there's still a small bug with some choices popping up a bit too much, I'll look into it.

I'm also experiencing with [a test suite](https://github.com/y-lohse/inkjs/blob/master/tests/ChoicesSpec.js) but I haven't gotten far with it yet.

A couple more code-related questions, if you don't mind:

1. In `VariableState.CopyFrom` ([here](https://github.com/inkle/ink/blob/master/ink-engine-runtime/VariablesState.cs#L94)), you're copying the global variables from another state. Is that a copy by reference or… not?
2. `VariablePointerValue` extends `StringValue`, which might cause a few problems in the js version, so I'd like to implement it as it's own class. Do you know how much it *really* needs from `StringValue`? Is it using the new line and white space properties and such?

**Comment by joethephish** [comment]
Awesome! Great to see it all working :-)

Cool to see you beginning work on a test suite... I guess the problem is that you can't directly copy my tests from the C# suite because they directly use source ink rather than the compiled JSON format :-/ I guess we need a new test suite that works on the Runtime.Object primitives at a lower level.

To answer your questions:

1. Urgh, so it's copying by reference, but it's relying on that not being a problem, since the variables' values should be treated as immutable (even if they're currently not strictly prevented from being mutated). Should probably fix that at some point!
2. Nope, it shouldn't need to be a `StringValue`, and isn't using newline / whitespace etc. It's purely just because it contains a string (the name of the variable being pointed to). It's probably semantically better to implement as its own thing anyway so it's not so misleading.

**Comment by joningold** [comment]
Wow, this is very exciting to see!

Playing through, I'm pretty certain you've got a bug with conditionals not
turning off options successfully. The easiest repro is to choose "think" on
the first turn - that should then get removed by the "once only" logic on
choices. But later choices with more variable-based logic are also showing
up when they shouldn't.

Still, it's clearly nearly there which is fantastic...!

cheers
jon

**Comment by y-lohse** [comment]
> I guess we need a new test suite that works on the Runtime.Object primitives at a lower level.

That would be ideal :-). I'm not overly happy with the format I currently have, because the strings are duplicated on the ink file and in the test. But I'm not sure if there's a way around it, probably not. I may also create a node wrapper around inklecate to be able to programmatically use the compiler, but a low level test suite would be better. 

Thanks for the other answers! To be honest, javascript isn't the best at enforcing immutability, so I'm kind of glad it's a copy by reference in this case. And regarding the choices, I haven't looked into stickies and fallbacks and such, so there's quite possibly a problem there. I'll investigate :-)

I'll keep you guys posted on further progress, thanks for the feedback!

**Comment by y-lohse** [comment]
Ok first the good news: the issue with choices was trivial, so that's fixed (haven't updated the demo though). I fixed a bunch of other things so by now, every language feature should be working. Which means I'm now working on the API side.

And here comes the bad news: I'm a bit stuck again :D. I'm trying to save & load a game state as JSON but I'm not sure what it's supposed to look like. Here's what I get for a simple hello world:

```
{"callstackThreads":{"threads":[{"callstack":[{"cPath":"","idx":0,"exp":false,"t
ype":0,"temp":{}}],"threadIndex":0}],"threadCounter":0},"variablesState":{},"eva
lStack":[],"outputStream":[],"currentChoices":[],"visitCounts":{},"turnIndices":
{},"turnIdx":-1,"storySeed":"1465157472488-8398","inkSaveVersion":2,"inkFormatVe
rsion":11}
```

Is the empty `cPath` normal? If I understand things correctly, it should reference the root container. However the empty string creates a Component which is neither named nor indexed, and subsequently can't be found when importing the JSON.

**Comment by joethephish** [comment]
Hrm, yeah, `cPath` shouldn't be empty I don't think. If anything, it should probably be missing, when between choices. At the point of serialisation [this check](https://github.com/inkle/ink/blob/master/ink-engine-runtime/CallStack.cs#L123) will often be false, since between choices, the current content pointer will have become null (since it's finished evaluating a piece of story). It's only between lines, i.e. between calls to `Continue` that it will be set to something valid, I believe.

I don't think the path should ever be an empty string, since there's no reason I can think of to reference the root container itself, unless it's some edge case where the story was saved after initialisation but before actually evaluating anything...?

**Comment by y-lohse** [comment]
> unless it's some edge case where the story was saved after initialisation but before actually evaluating anything...?

Thanks, that's totally what I was trying to do, only because it was an easy thing to test. After some debuging, saving and loading at choice points appears to be working now!

Unless I've missed something, here's what still needs to be done to get on par with the C# version:

- variables observers
- external functions
- catch up with the latest commits (Branches and such)

Nearly there!

**Comment by y-lohse** [comment]
Well, I think i'm mostly done :tada: 

Variable observers and external functions are working, and I've updated the demo and various packages. There are still a few minor things that I need to do, but nothing too important.

However, I haven't yet caught up with the changes to Branch.cs & co. These changes are currently on the dev branch of ink; should I wait until they land in master or can I go ahead and implement them?

**Comment by joethephish** [comment]
Woo congratulations! Looking forward to playing with it! Really appreciate
your efforts :)

Feel free to start implementing, we'll pull latest from dev into master
very soon.

**Comment by y-lohse** [comment]
I just published a new version with the changes from v0.3.5. I've also used the mighty 1.0 tag, so I think I implicitly pledged not to break inkjs inconsiderately :-).

There's [one known bug](https://github.com/y-lohse/inkjs/issues/11) but I suspect there are a few more lingering. I think we'll find them by using inkjs and writting more tests, though I probably won't work on that straight away.

So yeah, javascript port :white_check_mark:
